chat-with-tor/libp2p-demo-tutorial.md

```diff
@@ -0,0 +1,235 @@
+
+We will start with a few imports. These imports include `go-libp2p` itself, our pubsub library "floodsub", the IPFS DHT, and a few other helper packages to tie things together. 
+
+```go
+package main
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-ipfs-addr"
+	"github.com/libp2p/go-floodsub"
+	"github.com/libp2p/go-libp2p"
+	libp2pdht "github.com/libp2p/go-libp2p-kad-dht"
+	"github.com/libp2p/go-libp2p-peerstore"
+	"github.com/libp2p/go-libp2p-transport"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/multiformats/go-multihash"
+)
+```
+
+Next up, lets start constructing the pieces! First, we will set up our libp2p host. The host is the main abstraction that users of go-libp2p will deal with, It lets you connect to other peers, open new streams, and register protocol stream handlers.
+
+Next up, some configuration and initialization. We parse our command-line flags and start working on constructing our libp2p [`Host`](https://godoc.org/github.com/libp2p/go-libp2p-host#Host). We copy the user-provided addresses to listen on and optionally add a tor onion address to listen on if we're using the tor transport. **Note**: Due to some specifics of the tor transport layer's implementation, your local tor instance must be configured such that the SOCKS5 proxy and controller have the **same password**.
+
+```go
+func main() {
+	config := ParseFlags()
+	ctx := context.Background()
+
+	// Configure p2p host
+	var tpts []transport.Transport
+	addrs := make([]multiaddr.Multiaddr, len(config.ListenAddresses))
+	copy(addrs, config.ListenAddresses)
+
+	// Configure tor if options present
+	if config.TorConfig != nil {
+		torAddr, err := findOrCreateTorServiceAddress(config.TorConfig)
+		if err != nil {
+			panic(err)
+		}
+		addrs = append(addrs, torAddr)
+		torTpt, err := newOnionTransport(config.TorConfig)
+		if err != nil {
+			panic(err)
+		}
+		tpts = append(tpts, torTpt)
+
+		if err = addTorTransportToAddrUtil(); err != nil {
+			panic(err)
+		}
+	}
+```
+
+With our new transport configured (in the event that we're using tor), we can create our `Host`. The `Host` object will be most developer's primary point-of-contact with `libp2p`. The `Host` is so-named because it is both a client and a server. Behind the scenes, it manages a whole host (groan) of difficult tasks, including connection management and reuse, stream multiplexing, and encryption. After creating our host, we print out all of the addresses at which it can be reached so that we might tell other clients. This subtly showcases some of `libp2p`'s power, potentially many transports and protocols supported by a single, simple abstraction.
+
+```go
+tptOptions := libp2p.Transports(tpts...)
```

Comment:
FYI, multiple transport/listen options will be merged into one. You can actually just declare an "options" slice and start appending random options to it (transports, listen addresses, etc).

However, this also works.